lib/features/build_server/cache/project_cache.dart

```diff
@@ -0,0 +1,235 @@
+      }
+    }
+
+    return required;
+  }
+
+  /// Persists the passed [file].
+  Future<void> insert(SourceFile file) async {
+    final path = joinAll([
+      _workingDirectory.path,
+      'src',
+      '${file.entry.toString()}',
+      ...file.path,
+      '${file.name}.${file.extension}'
+    ]);
+    final diskFile = File(path);
+
+    await diskFile.create(recursive: true);
+    await diskFile.writeAsBytes(await file.data, flush: true);
+
+    if (_lock.files.where((f) => f.id == file.id).isNotEmpty) {
+      final lockFile = _lock.files.singleWhere((f) => f.id == file.id);
+      lockFile.hash = await file.hash;
+    } else {
+      _lock.files.add(SourceFileLock(id: file.id, hash: await file.hash, path: path));
+    }
+  }
+
+  /// A list of all stored header files.
+  List<File> get headerFiles {
+    final headers = <File>[];
+
+    for (final file in _lock.files) {
+      final ext = extension(file.path);
+
+      if (FileExtension.headers.contains(ext)) {
+        headers.add(File(file.path));
+      }
+    }
+
+    return headers;
+  }
+
+  /// A list of all stored source files.
+  ///
+  /// Each source file is returned as an object which contains
+  /// a path to the source file itself and to the object file.
+  /// The compiler must use this path to create the corresponding object file.
+  /// After compilation the `link` method has to be called, to let
+  /// the cache know that the latest version of the file has been compiled.
+  /// ```dart
+  /// final refs = await cache.sourceFiles;
+  ///
+  /// for (final ref in refs) {
+  ///   await compile(sourcePath: ref.source, objectPath: ref.object);
+  ///   await ref.link();
+  /// }
+  /// ```
+  Future<List<ImplementationObjectRef>> get sourceFiles async {
+    final refs = <ImplementationObjectRef>[];
+
+    for (final file in _lock.files) {
+      if (_isCompiled(file.id)) {
+        continue;
+      }
+
+      final ext = extension(file.path);
+
+      if (FileExtension.sources.contains(ext)) {
+        final sourceFile = File(file.path);
+        final objectFile = File(_objectPath(file.id));
+
+        await objectFile.create(recursive: true);
+
+        refs.add(ImplementationObjectRef(sourceFile, objectFile, () async {
+          final exists = await objectFile.exists();
+          final isNotEmpty = (await objectFile.stat()).size > 0;
+
+          if (exists && isNotEmpty) {
+            _setIsCompiled(file.id, true);
+          }
+        }));
+      }
+    }
+
+    return refs;
+  }
+
+  /// A list of all stored object files.
+  /// Only those are returned which represent the latest
+  /// version of the associated source file.
+  Future<List<File>> get objectFiles async {
+    final objectFiles = <File>[];
+
+    for (final file in _lock.files) {
+      if (_isCompiled(file.id)) {
+        objectFiles.add(File(_objectPath(file.id)));
+      }
+    }
+
+    return objectFiles;
+  }
+
+  /// The path of the executable.
```

Comment:
Same thing as with object files. The method returns a `File` object instead of raw data or something. That's better for memory and serves the purpose of a cache well enough. Maybe it should check whether the file exists and throw an error if not. But everything else is the job of a higher order object like the build server.

Reply:
I didn't mean to pass the files data, but this could be confusing (like you said) when the executable is not yet compiled.

Reply:
Ok I'll throw an error if no executable exists

Reply:
Order back! No error. The executable path provided by the cache is also required for g++ to put the said executable exactly there. Use `Compiler.latestExecutable` to receive `null` if no runnable binary is available